### PR TITLE
server: test server

### PR DESCRIPTION
**Pull request by Tricky511**
Add `/proto/engine.proto` and `/server/src/distributed_engine/engine_rpc` to implement servers communication

**Add test**

server: test the distributed engine by script

fix: delete_dir function use directory_delete_entry, instead of directory_add_entry

fix: use hash get address, instead of index

directory empty implemented in delete_directory function

add some case for OperationType

add some case for EngineErr2Status temporarily

**Test Steps:**

1. You need to run manager
2. You need to run three different servers, you need to manually replace the configuration file and start it three times (refer to the three configuration files in the test_yaml directory)
3. run server_test.py to test

### CODE REVIEW

File: server/src/distributed_engine/mod.rs [hunk 1]
```diff
@@ -1,62 +1,259 @@
...
+                        std::io::ErrorKind::NotConnected,
+                    )))
+                }
+            };
+            if status != 0 {
+                return into_engineerror(status);
+            }
+        }
+
+        self.local_storage.final_create_directory(path)
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:93]: why need to rename?

File: server/src/distributed_engine/mod.rs [hunk 2]
```diff
@@ -1,62 +1,259 @@
...
+        }
+        if self.local_storage.is_exist(path.clone())? {
+            return Err(EngineError::Exist);
+        }
+
+        let index = match path.rfind('/') {
+            Some(value) => value,
+            None => return Err(EngineError::Path),
+        };
+        let (parent_dir, file_name) = (&path[..=index], &path[(index + 1)..]);
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:161]: abstract to a function

File: server/src/distributed_engine/mod.rs [hunk 3]
```diff
@@ -1,62 +1,259 @@
...
+
+        let index = self.get_connection_index(parent_dir).unwrap();
+        if self.index == index {
+            self.local_storage
+                .directory_add_file(parent_dir.to_string(), file_name.to_string())?;
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:166]: `directory_add_entry` might be better

File: server/src/distributed_engine/engine_rpc.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            .directory_delete_file(message.parentdir.clone(), message.filename.clone());
+
+        let response = EngineResponse {
+            status: into_status(result),
+        };
+        Ok(Response::new(response))
+    }
+}
+
+pub fn into_status(result: Result<(), EngineError>) -> u32 {
```
Comment (uran0sH) [inline server/src/distributed_engine/engine_rpc.rs:63]: use `From` trait

File: server/src/distributed_engine/mod.rs [hunk 5]
```diff
@@ -1,62 +1,259 @@
...
+        let index = match path.rfind('/') {
+            Some(value) => value,
+            None => return Err(EngineError::Path),
+        };
+        let (parent_dir, file_name) = (&path[..=index], &path[(index + 1)..]);
+
+        let index = self.get_connection_index(parent_dir).unwrap();
+        if self.index == index {
+            self.local_storage
+                .directory_delete_file(parent_dir.to_string(), file_name.to_string())?;
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:214]: file -> entry

File: server/src/distributed_engine/mod.rs [hunk 6]
```diff
@@ -1,62 +1,259 @@
...
+                        std::io::ErrorKind::NotConnected,
+                    )))
+                }
+            };
+            if status != 0 {
+                return into_engineerror(status);
+            }
+        }
+
+        self.local_storage.final_delete_file(path)
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:242]: don't need to rename

File: server/src/storage_engine/default_engine.rs [hunk 7]
```diff
@@ -45,7 +45,7 @@ impl StorageEngine for DefaultEngine {
             let mut db_opts = Options::default();
             db_opts.create_if_missing(true);
             let path = format!("{}_dir", db_path);
-            let db = match DB::open(&db_opts, path.as_str()) {
+             let db = match DB::open(&db_opts, path.as_str()) {
```
Comment (uran0sH) [inline server/src/storage_engine/default_engine.rs:48]: remove space

File: server/src/storage_engine/default_engine.rs [hunk 8]
```diff
@@ -237,16 +156,59 @@ impl StorageEngine for DefaultEngine {
...
+        self.delete_from_parent(path)?;
+        Ok(())
+    }
 
-        let index = match path.rfind('/') {
-            Some(value) => value,
-            None => return Err(EngineError::Path),
-        };
-        let (parent_dir, file_name) = (&path[..=index], &path[(index + 1)..]);
+    fn is_exist(&self, path: String) -> Result<bool, EngineError> {
```
Comment (uran0sH) [inline server/src/storage_engine/default_engine.rs:166]: Returning `Result<(), EngineError>` might be enough.

File: server/src/storage_engine/mod.rs [hunk 9]
```diff
@@ -22,9 +18,25 @@ pub trait StorageEngine {
...
     fn write_file(&self, path: String, data: &[u8]) -> Result<(), EngineError>;
 
-    fn delete_file(&self, path: String) -> Result<(), EngineError>;
+    fn delete_directory_recursive(&self, path: String) -> Result<(), EngineError>;
 
-    fn delete_directory(&self, path: String) -> Result<(), EngineError>;
+    fn is_exist(&self, path: String) -> Result<bool, EngineError>;
 
-    fn delete_directory_recursive(&self, path: String) -> Result<(), EngineError>;
+    fn directory_empty(&self, path: String) -> Result<(), EngineError>;
```
Comment (uran0sH) [inline server/src/storage_engine/mod.rs:25]: it's a good one. But it doesn't need to add to this trait.

File: server/src/storage_engine/mod.rs [hunk 10]
```diff
@@ -22,9 +18,25 @@ pub trait StorageEngine {
...
+        file_name: String,
+    ) -> Result<(), EngineError>;
+
+    fn final_create_file(&self, path: String) -> Result<(), EngineError>;
+
+    fn final_delete_file(&self, path: String) -> Result<(), EngineError>;
+
+    fn final_create_directory(&self, path: String) -> Result<(), EngineError>;
+
+    fn final_delete_directory(&self, path: String) -> Result<(), EngineError>;
```
Comment (uran0sH) [inline server/src/storage_engine/mod.rs:41]: don't need to rename

**Review by uran0sH — CHANGES_REQUESTED** [review]
Plz modify your commit title. I can't get any information from your commit title. Please clarify what you do in this commit, and elaborate on how you do in the commit message.

And I think we should use a custom protocol instead of using tonic.

File: server/src/storage_engine/default_engine.rs (hunk 8 above)
Comment (Tricky511) [reply server/src/storage_engine/default_engine.rs:166]: The `Result` of this function has different handling when used by `create` and `delete`, I don't think it's good to return `EngineError` directly

Comment (uran0sH) [reply server/src/storage_engine/default_engine.rs:166]: fine

File: server/src/distributed_engine/mod.rs (hunk 1 above)
Comment (Tricky511) [reply server/src/distributed_engine/mod.rs]: done

File: server/src/distributed_engine/mod.rs (hunk 2 above)
Comment (Tricky511) [reply server/src/distributed_engine/mod.rs]: done

File: server/src/distributed_engine/mod.rs (hunk 3 above)
Comment (Tricky511) [reply server/src/distributed_engine/mod.rs]: done

File: server/src/distributed_engine/engine_rpc.rs (hunk 4 above)
Comment (Tricky511) [reply server/src/distributed_engine/engine_rpc.rs]: done

File: server/src/storage_engine/mod.rs (hunk 10 above)
Comment (Tricky511) [reply server/src/storage_engine/mod.rs]: done

File: server/src/storage_engine/mod.rs (hunk 9 above)
Comment (Tricky511) [reply server/src/storage_engine/mod.rs]: done

File: server/src/storage_engine/default_engine.rs (hunk 7 above)
Comment (Tricky511) [reply server/src/storage_engine/default_engine.rs]: done

File: server/src/distributed_engine/mod.rs (hunk 5 above)
Comment (Tricky511) [reply server/src/distributed_engine/mod.rs]: done

File: server/src/distributed_engine/mod.rs (hunk 6 above)
Comment (Tricky511) [reply server/src/distributed_engine/mod.rs]: done

**Comment by uran0sH** [comment]
'implement distributed engine with tonic' may be a better title.

File: server/src/distributed_engine/mod.rs [hunk 11]
```diff
@@ -1,62 +1,258 @@
...
+                    .get(&index)
+                    .unwrap()
+                    .clone()
+            };
+            let response = connect.directory_add_entry(request).await;
+
+            let status = match response {
+                Ok(value) => value.get_ref().status,
+                _ => {
+                    return Err(EngineError::StdIo(std::io::Error::from(
```
Comment (luanshaotong) [inline server/src/distributed_engine/mod.rs:94]: Should we release the lock in the local database？

**Review by luanshaotong — APPROVED** [review]
need test

**Comment by luanshaotong** [comment]
> Plz modify your commit title. I can't get any information from your commit title. Please clarify what you do in this commit, and elaborate on how you do in the commit message.
> 
> And I think we should use a custom protocol instead of using tonic.

Our protocol isn’t well tested and I think it‘s OK to use tonic for now.

File: server/src/distributed_engine/mod.rs [hunk 12]
```diff
@@ -1,62 +1,258 @@
...
+            });
+            let mut connect = {
+                self.connections
+                    .lock()
+                    .unwrap()
+                    .get(&index)
+                    .unwrap()
+                    .clone()
+            };
+            let response = connect.directory_add_entry(request).await;
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:133]: `directory_delete_entry`

File: server/src/server.rs [hunk 13]
```diff
@@ -20,19 +20,21 @@ use crate::{distributed_engine::DistributedEngine, storage_engine::StorageEngine
...
+            EngineError::IO => libc::EIO,
+            EngineError::NoEntry => libc::ENOENT,
+            EngineError::NotDir => libc::ENOTDIR,
+            EngineError::IsDir => libc::EISDIR,
+            EngineError::Exist => libc::EEXIST,
+            EngineError::NotEmpty => libc::ENOTEMPTY,
+            // temp
+            EngineError::Path => -1,
+            EngineError::StdIo(_) => -1,
+            EngineError::Rocksdb(_) => -1,
```
Comment (luanshaotong) [inline server/src/server.rs:32]: Why the error codes are all -1?

Comment (luanshaotong) [reply server/src/server.rs:32]: Maybe we can use numbers >1000?

Comment (uran0sH) [reply server/src/server.rs:32]: I think we don't specify these values now.

File: server/src/distributed_engine/mod.rs [hunk 14]
```diff
@@ -1,62 +1,263 @@
...
+use engine_rpc::enginerpc::{enginerpc_client::EnginerpcClient, EngineRequest};
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use tonic::transport::Channel;
 
 pub struct DistributedEngine<Storage: StorageEngine> {
-    pub local_storage: Storage,
+    pub address: String,
+    pub local_storage: Arc<Storage>,
+    pub connections: Arc<Mutex<HashMap<String, Option<EnginerpcClient<Channel>>>>>,
```
Comment (uran0sH) [inline server/src/distributed_engine/mod.rs:15]: We might use `dashmap` first.

**Review by uran0sH — CHANGES_REQUESTED** [review]

File: server/src/storage_engine/default_engine.rs [hunk 15]
```diff
@@ -394,12 +362,14 @@ mod tests {
     use crate::storage_engine::{default_engine::generate_local_file_name, StorageEngine};
 
     use super::{DefaultEngine, SubDirectory};
-
```
Comment (uran0sH) [inline server/src/storage_engine/default_engine.rs:397]: don't modify here

File: server/src/distributed_engine/mod.rs (hunk 14 above)
Comment (luanshaotong) [reply server/src/distributed_engine/mod.rs:15]: agree

**Comment by uran0sH** [comment]
This PR's test should test Distributed Engine, not test the whole server.

**Pull request closed without merging**